sys/congure/quic/congure_quic.c

```diff
@@ -0,0 +1,289 @@
+/*
+ * Copyright (C) 2021 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @{
+ *
+ * @file
+ * @author  Martine Lenders <m.lenders@fu-berlin.de>
+ *
+ * See https://tools.ietf.org/html/draft-ietf-quic-recovery-34#appendix-B and
+ * parts of https://tools.ietf.org/html/draft-ietf-quic-recovery-34#appendix-A
+ * (for pacing calculation) as basis for this implementation.
+ */
+
+#include <assert.h>
+#include <limits.h>
+#include <stdint.h>
+
+#include "clist.h"
+#include "timex.h"
+#include "ztimer.h"
+
+#include "congure/quic.h"
+
+static void _snd_init(congure_snd_t *cong, void *ctx);
+static int32_t _snd_inter_msg_interval(congure_snd_t *cong, unsigned msg_size);
+static void _snd_report_msg_sent(congure_snd_t *cong, unsigned sent_size);
+static void _snd_report_msg_discarded(congure_snd_t *cong, unsigned msg_size);
+static void _snd_report_msgs_lost(congure_snd_t *cong, congure_snd_msg_t *msgs);
+static void _snd_report_msg_acked(congure_snd_t *cong, congure_snd_msg_t *msg,
+                                  congure_snd_ack_t *ack);
+static void _snd_report_ecn_ce(congure_snd_t *cong, ztimer_now_t time);
+
+static const congure_snd_driver_t _driver = {
+    .init = _snd_init,
+    .inter_msg_interval = _snd_inter_msg_interval,
+    .report_msg_sent = _snd_report_msg_sent,
+    .report_msg_discarded = _snd_report_msg_discarded,
+    .report_msgs_timeout = _snd_report_msgs_lost,
+    .report_msgs_lost = _snd_report_msgs_lost,
+    .report_msg_acked = _snd_report_msg_acked,
+    .report_ecn_ce = _snd_report_ecn_ce,
+};
+
+static inline bool _in_recov(congure_quic_snd_t *c, ztimer_now_t sent_time)
+{
+    return sent_time <= c->recovery_start;
+}
+
+static void _on_congestion_event(congure_quic_snd_t *c, ztimer_now_t sent_time)
+{
+    if (_in_recov(c, sent_time)) {
+        return;
+    }
+    /* enter congestion recovery period */
+    c->recovery_start = ztimer_now(ZTIMER_MSEC);
+    c->ssthresh = (c->super.cwnd * c->consts->loss_reduction_numerator) /
+                  c->consts->loss_reduction_denominator;
```

Comment:
Feel free to ignore, but doesn't
```suggestion
    c->ssthresh = (c->super.cwnd * c->consts->loss_reduction_numerator)
                /  c->consts->loss_reduction_denominator;
```
look nicer?